geniusyield-market-maker/src/GeniusYield/MarketMaker/MakerBot.hs

```diff
@@ -115,19 +130,129 @@ buildAndSubmitActions user@User {uColl, uStakeCred} providers netId ua di = flip
+                       lift $ gyLogWarning providers logNS $ "Outrageous price mismatch among providers!"
+                       put MBSpooked2 { mbs2Relax = 0 }
+
+                     PriceUnavailable  → do
+                       lift $ gyLogWarning providers logNS $ "All Prices Providers unavailable!"
+                       put MBSpooked2 { mbs2Relax = 0 }
+
+                     psf@PriceSourceFail {} → do
+                       lift $ logPricesProviderFail providers psf
+
+                     _                 → do
+                       lift $ gyLogInfo providers logNS $ "Apparent recovery of Prices Providers; waiting for observation period to elapse."
+                       put MBSpooked1 { mbs1Relax = mbs1Relax + 1, mbs1Worse = mbs1Worse }
+
+                   lift $ threadDelay respiteDelay
+
+    MBSpooked2 {..} → do
+      if mbs2Relax > pccAfterExitRelaxAim2 cfg then do
+           lift $ gyLogInfo providers logNS $ "Resuming strategy"
+           put MBReady
+         else do
+           pe ← lift $ priceEstimate' pp mbToken
+
+           case pe of
+             PriceMismatch1    → do
+               lift $ gyLogWarning providers logNS $ "Price mismatch persists"
+               put MBSpooked2 { mbs2Relax = 0 }
+
+             PriceMismatch2    → do
+               lift $ gyLogWarning providers logNS $ "Outrageous price mismatch persists"
+               put MBSpooked2 { mbs2Relax = 0 }
+
+             PriceUnavailable  → do
+               lift $ gyLogWarning providers logNS $ "All Prices Providers unavailable"
+               put MBSpooked2 { mbs2Relax = 0 }
+
+             psf@PriceSourceFail {} → do
+               lift $ logPricesProviderFail providers psf
+
+             _                 → do
+               lift $ gyLogInfo providers logNS $ "Apparent recovery of Prices Providers; waiting for observation period to elapse."
+               put MBSpooked2 { mbs2Relax = mbs2Relax + 1 }
+
+           lift $ threadDelay respiteDelay
+
+evolveStrategy
+  ∷ Strategy
+  → MakerBot
+  → GYNetworkId
+  → GYProviders
+  → PricesProviders
+  → DEXInfo
+  → StateT MBFret IO ()
+evolveStrategy runStrategy mb netId providers pp di =
+  forever $ get >>= mbStateMachine runStrategy mb netId providers pp di
+        
 executeStrategy
-  :: Strategy
-  -> MakerBot
-  -> GYNetworkId
-  -> GYProviders
-  -> PricesProviders
-  -> DEXInfo
-  -> IO ()
-executeStrategy runStrategy MakerBot {mbUser, mbDelay, mbToken} netId providers pp di =
-  forever $ do
-    newActions ← runStrategy pp mbUser mbToken
-
-    buildAndSubmitActions mbUser providers netId newActions di
-
-    gyLogInfo providers logNS "---------- Done for the block! -----------"
-    threadDelay mbDelay
+  ∷ Strategy
+  → MakerBot
+  → GYNetworkId
+  → GYProviders
+  → PricesProviders
+  → DEXInfo
+  → IO ()
+executeStrategy runStrategy mb netId providers pp di = do
+  runStateT
+    (evolveStrategy runStrategy mb netId providers pp di)
+    MBReady
+  return ()
```

Comment:
Is `return ()` here needed?

Reply:
Done.

Reply:
In general, `void $ ...` is helpful in such scenarios.